chore(reports): add task for slack channels warm-up

Usiel opened the pull request:
### SUMMARY

Adds a celery task to warm up the Slack channel cache and allows admins to schedule the same task (with example). This is helpful for cases where retrieval of the full channel list takes minutes.

#### Changes to the `get_channels` function

The Slack conversations API is not necessarily faster when using the `types` filter: For example, when we filter for private channels, the cursor will still iterate over all public channels and the number of requests remains the same, only the data transmitted is less.

With this in mind, I believe it is more efficient for us to cache only a single value for `get_channels()` and avoid having multiple cache values for each possible value of `types`. Doing the filtering in `get_channels_with_search` on demand doesn't add much cost.

Additionally, the `limit` param is also not relevant for the cache key and is inlined.

... and, of course, this change makes the celery task more sensible :)

#### Warning

There is a minor API change in this PR: Previously, we followed Slack's [conversations.list](https://api.slack.com/methods/conversations.list) convention, where empty an `types` param means that only public channels are returned. After this PR we will return all channel types when `types` is empty or null. To me that feels more logical, but if people feel we should stick to previous approach I'm OK with changing this.

### TESTING INSTRUCTIONS

Pre:
- Set `SLACK_API_TOKEN`
- Enable `ALERT_REPORTS`

#### Test Celery Task

```
celery --app=superset.tasks.celery_app:app call slack.cache_channels
```

**Expected**: Task is launched and result is cached.

#### Test Beat Schedule Example

1. Uncomment the beat schedule entry for `slack.cache_channels`
2. [Optional]: Adapt schedule to trigger every minute (or set time)

**Expected**: Task is scheduled and result is cached.

#### Test REST API

Try for `TYPES={"public_channel", "private_channel", "", "public_channel, private_channel"}`

```
curl -X 'GET' \
  'http://localhost:8088/api/v1/report/slack_channels/?q=(types%3A!(<TYPES>))' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer <TOKEN>'
```

**Expected**: Cached value is used, filtered and returned.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: https://github.com/apache/superset/issues/32480
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

cc @Vitor-Avila since this builds on your recent work in https://github.com/apache/superset/pull/32529

Usiel committed:
Cache entire channel list and do filtering on request

The Slack conversations API is not necessarily faster when using the `types` filter: For example,
when we filter for private channels, the cursor will still iterate over all public channels and the
number of requests remains the same, only the data transmitted is less.

With this in mind, I believe it is more efficient for us to cache only a single value for `get_channels()`
and avoid having multiple cache values for each possible value of `types`. Doing the filtering in
`get_channels_with_search` on demand doesn't add much cost.

Additionally, the `limit` param is also not relevant for the cache key and is inlined.

Usiel committed:
Adds cache warm-up job for Slack channels

Adds a celery task to warm up the Slack channel cache and allows admins to schedule
the same task (with example). This is helpful for cases where retrieval of the full
channel list takes minutes.

### Testing

Pre:
- Set `SLACK_API_TOKEN`
- Enable `ALERT_REPORTS`

#### Test Celery Task

```
celery --app=superset.tasks.celery_app:app call slack.cache_channels
```

**Expected**: Task is launched and result is cached.

#### Test Beat Schedule Example

1. Uncomment the beat schedule entry for `slack.cache_channels`
2. [Optional]: Adapt schedule to trigger every minute (or set time)

**Expected**: Task is scheduled and result is cached.

#### Test REST API

Try for `TYPES={"public_channel", "private_channel", "", "public_channel, private_channel"}

```
curl -X 'GET' \
  'http://localhost:8088/api/v1/report/slack_channels/?q=(types%3A!(<TYPES>))' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer <TOKEN>'
```

**Expected**: Cached value is filtered and returned.

Usiel commented on superset/utils/slack.py:
```diff
@@ -99,18 +99,25 @@ def get_channels_with_search(
         raise SupersetException(f"Failed to list channels: {ex}") from ex
 
+    if types and not len(types) == len(SlackChannelTypes):
+        conditions: list[Callable[[SlackChannelSchema], bool]] = []
```
The intention here was to make this more easily pluggable in the future if we add more types.

Vitor-Avila replied:
Nice!

Usiel commented on superset/tasks/slack.py:
```diff
@@ -0,0 +1,27 @@
+logger = logging.getLogger(__name__)
+
+
+@celery_app.task(name="slack.cache_channels")
+def cache_channels() -> None:
+    get_channels(force=True)
```
The Slack client handles both connection and (because we added it) rate-limit errors.

Usiel commented on superset/utils/slack.py:
```diff
@@ -99,18 +99,25 @@ def get_channels_with_search(
     try:
         channels = get_channels(
-            limit=limit,
-            extra_params=extra_params,
             force=force,
             cache_timeout=86400,
         )
```
The parameters are from utils.cache.memoized_func

Vitor-Avila commented on superset/config.py:
```diff
@@ -1046,6 +1047,11 @@ class CeleryConfig:  # pylint: disable=too-few-public-methods
         #     "schedule": crontab(minute="*", hour="*"),
         #     "kwargs": {"retention_period_days": 180},
         # },
+        # Uncomment to enable Slack channel cache warm-up
+        # "slack.cache_channels": {
+        #     "task": "slack.cache_channels",
+        #     "schedule": crontab(minute="0", hour="*"),
+        # },
```
Nice!

Vitor-Avila commented on superset/tasks/slack.py:
```diff
@@ -0,0 +1,27 @@
+
+@celery_app.task(name="slack.cache_channels")
+def cache_channels() -> None:
+    get_channels(force=True)
```
Could we enforce calling it with a `cache_timeout` as well? Just to prevent creating cache items that don't have an expiration (since it's a scheduled task it would always refresh on schedule, but there's an edge case of a user disabling the task and then it would be there until someone refreshes through the UI which has a `cache_timeout`).

I was actually wondering if we should have a new config like `SLACK_CACHE_TIMEOUT` (since this is hardcoded in the code, keeping it in `config.py` might make more sense). This was supposed to be part of my PR, but just thought about it after having it merged so was planning to do a follow up in the future. If we have a config, we could use it here + in the API config.

Vitor-Avila replied:
Also, since this is calling `get_channels()` directly, could we do it in a `try/except` block and log exceptions? Raising an error wouldn't make much sense as this is async, but I think logging is always a good thing.

Usiel replied:
Makes sense! Done and done.

sadpandajoe commented on superset/utils/slack.py:
```diff
@@ -71,11 +71,12 @@ def get_channels(limit: int, extra_params: dict[str, Any]) -> list[SlackChannelS
     while True:
         response = client.conversations_list(
-            limit=limit, cursor=cursor, exclude_archived=True, **extra_params
+            limit=999, cursor=cursor, exclude_archived=True, **extra_params
```
why are we hardcoding this now?

Usiel replied:
I inlined the `limit` because it never governed the amount of channels returned by the function. The function always returns all channels. The `limit` parameter here only governs the amount of channels returned by each Slack API response. 

Going for the maximum (999) allowed by Slack is most efficient, imo. Lower values are more likely to result in rate-limiting (which is based on requests, not rows returned or similar).

Usiel replied:
If we see a need we could add paging to our `/slack_channels` endpoint (and `get_channels_with_search(...)`). I don't know about that though, our only usage is to get all channels to give the user a choice for the report notification setting.

Usiel committed:
Adds new SLACK_CACHE_TIMEOUT and makes use of it

The new configuration field defaults to 1 day. Currently, the only cache affected is the Slack channels cache.

Usiel committed:
Adds logging to Slack cache channel task

Vitor-Avila reviewed (approved):
LGTM!

eschutho commented on superset/tasks/slack.py:
```diff
@@ -0,0 +1,35 @@
+def cache_channels() -> None:
+    try:
+        get_channels(
+            force=True, cache_timeout=current_app.config["SLACK_CACHE_TIMEOUT"]
```
@dpgaspar can you confirm if this task would work properly for fetching the slack token (when it is a callable) for multi-tenant applications? Do we need to pass in some user context to this function?

dpgaspar replied:
it should just work

rusackas commented:
Looks good to merge?

Vitor-Avila commented:
@rusackas my only question here is relating to the API change. While this API endpoint is solely used by the Slack integration (I can't think why a user would create any workflow on top of a Superset API endpoint that just lists Slack channels) I wonder if we need to mention this in `UPDATING.md` and if it has to be considered a breaking change or not.
